ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/Committer.scala

```diff
@@ -78,46 +87,63 @@ private[committer] trait Committer[PartialResult] extends SubmissionExecutor {
       logEntry -> ctx.getOutputs.toMap
     }
 
-  def dryRun(
+  def runWithPreExecution(
       submission: DamlSubmission,
       participantId: ParticipantId,
       inputState: DamlStateMapWithFingerprints,
-  ): PreexecutionResult = {
-    runTimer.time { () =>
-      // TODO(miklos): Create context for pre-execution here.
+  ): PreExecutionResult = {
+    preExecutionRunTimer.time { () =>
       val commitContext = new CommitContext {
         override def getRecordTime: Option[Time.Timestamp] = None
 
         override def getParticipantId: ParticipantId = participantId
 
         override def inputsWithFingerprints: DamlStateMapWithFingerprints = inputState
       }
-      preexecute(submission, participantId, inputState, commitContext)
+      preExecute(submission, participantId, inputState, commitContext)
     }
   }
 
-  private[committer] def preexecute(
+  private[committer] def preExecute(
       submission: DamlSubmission,
       participantId: ParticipantId,
       inputState: DamlStateMapWithFingerprints,
       commitContext: CommitContext,
-  ): PreexecutionResult = {
+  ): PreExecutionResult = {
     val logEntry = runSteps(commitContext, submission)
-    PreexecutionResult(
+    PreExecutionResult(
       readSet = commitContext.getAccessedInputKeysWithFingerprints.toMap,
       successfulLogEntry = logEntry,
       stateUpdates = commitContext.getOutputs.toMap,
-      // TODO(miklos): Take out-of-time-bounds log entry and min/max record time from context.
       outOfTimeBoundsLogEntry = constructOutOfTimeBoundsLogEntry(commitContext),
-      minimumRecordTime = Timestamp.MinValue,
-      maximumRecordTime = Timestamp.MaxValue,
+      minimumRecordTime = commitContext.minimumRecordTime
+        .map(Timestamp.assertFromInstant)
+        .getOrElse(Timestamp.MinValue),
+      maximumRecordTime = commitContext.maximumRecordTime
+        .map(Timestamp.assertFromInstant)
+        .getOrElse(Timestamp.MaxValue),
       // We assume the time updates must be visible to every participant.
       involvedParticipants = AllParticipants
     )
   }
 
   private def constructOutOfTimeBoundsLogEntry(commitContext: CommitContext): DamlLogEntry =
-    DamlLogEntry.getDefaultInstance
+    commitContext.outOfTimeBoundsLogEntry
+      .map { rejectionLogEntry =>
+        val builder = DamlOutOfTimeBoundsEntry.newBuilder
+          .setEntry(rejectionLogEntry)
+        commitContext.minimumRecordTime.foreach { instant =>
+          builder.setTooEarlyUntil(buildTimestamp(instant))
+        }
+        commitContext.maximumRecordTime.foreach { instant =>
+          builder.setTooLateFrom(buildTimestamp(instant))
+        }
+        DamlLogEntry.newBuilder
+          .setOutOfTimeBoundsEntry(builder)
+          .build
```

Comment:
Where is `deduplicate_until` set in the `OutOfTimeBoundsEntry`?

Reply:
See #6672 which builds on this one.